KKUMI-16 get banners

eekrwl opened the pull request:
## 구현내용
### 1. 배너 하나 조회 api 구현
홈화면 상단에서 호출될 배너를 전체 조회하는 api를 구현했습니다.
#### 스프링 데이터 JPA 사용
>아래와 같이 스프링 데이터 JPA를 사용하였습니다.
>```java
>public interface BannerRepository extends JpaRepository<Banner, Long> {
>}
>.```
#### 엔티티에 detailImageUrl 추가
>처음에는 배너 엔티티에 아래 두개의 컬럼만 존재해서 간단히 만들었습니다. 
>```
>    private Long id;
>    private String imageUrl;
>```
>하지만 생각해보니 홈화면에서 보이는 배너는 요약된 대표화면이고, 클릭을 했을때 보이는 상세 이미지가 없다는 생각이 들었습니다. 그래>서 아래와 같이 detailImageUrl을 추가했습니다.
>```
>    private Long id;
>    private String imageUrl;
>    private String detailImageUrl;
>```
#### BannerService
>BannerService에서 리포지토리에 접근하여 banner_id로 디테일이미지를 찾아주는 부분을 구현했습니다. 
>```java
>    public BannerDto getBannerById(Long bannerId) {
>        Banner banner = bannerRepository.findById(bannerId)
>                .orElseThrow(() -> new CommonException(ErrorCode.BANNER_NOT_FOUND,"존재하지 않는 배너입니다.","해당 id의 배너를 찾을 수 없습니다."));
>        return BannerDto.builder()
>                .id(banner.getId())
>                .imageUrl(banner.getDetailImageUrl())
>                .build();
>    }
>```
>BannerDto를 만들어 사용했습니다.
>```java
>    public class BannerDto {
>        private Long id;
>        private String imageUrl;
>    }
>```
   
     
### 2. 배너 전체 조회 api 구현
홈화면 상단에서 보여줄 배너 전체를 조회하는 api를 구현했습니다.
하나 조회와 다르게, 상세이미지 대신 일반 이미지를 조회해주었습니다.
#### BannerService
>아래는 BannerService에서 구현한 부분인데, 코드가 길어지기에 getSimpleBanners() 메서드를 따로 분리했습니다.
>Banner엔티티에는 imageUrl과 detailImageUrl이 함께 있는데, 여기서 필요한 부분은 detailImageUrl이라 고민하다가 위에서 사용했던 `BannerDto를 다시 사용`해줬습니다.
>```java
>    public BannerListResponse getAllBanners() {
>        List<BannerDto> bannerList = getSimpleBanners();
>        return BannerListResponse.builder()
>                .banners(bannerList)
>                .build();
>    }
>
>     private List<BannerDto> getSimpleBanners() {
>        List<Banner> bannerList = bannerRepository.findAll();
>        return bannerList.stream()
>                .map(banner -> BannerDto.builder()
>                        .id(banner.getId())
>                        .imageUrl(banner.getImageUrl())
>                        .build())
>                .collect(Collectors.toList());
>    }
>```      
    
## 고민사항
### 1. 배너 전체 조회 api에서 메서드 두개로 분리해줬는데 하나에 해도 되는걸 괜히 쪼갰는지 고민중입니다.
하나에 다 넣어도 될 것 같은데 보통 어느정도 길이가 넘어가면 분리하는지 궁금합니다.
### 2. 예외의 크기가 적당한지 고민입니다.
```java
new CommonException(ErrorCode.BANNER_NOT_FOUND,"존재하지 않는 배너입니다.","해당 id의 배너를 찾을 수 없습니다.")
```
id로 배너 조회시 해당 id 배너가 존재하지 않는다면 위와 같이 예외처리를 했습니다. ErrorCode에 BANNER_NOT_FOUND라고 넣어놨는데, 크기를 너무 작게 잡아 재사용할 일이 없는 예외를 만든건지 고민입니다. 그냥 해당 자료가 없을 때, ErrorCode에는 NOT_FOUND 정도만 넣어놓고, 상세 설명에서 배너가 존재하지 않는다고 적어주는 것으로 바꿀지 고민입니다.
### 3. aws secrets 관리
aws secrets 관리를 위해 aws secrets manager를 사용했습니다.
예전에는 aws.yml과 같은 파일에 secrtes를 넣어두고 .gitignore에 해당 파일을 넣어 깃허브에 올라가지 않도록 했었습니다.
이번에는 시크릿 관리에 편리할 것 같아 secrets manager를 사용해봤습니다. 
**dev/mykkumi/s3**
![스크린샷 2024-06-26 오후 1 17 03](https://github.com/SW-Marastro/Mykkumi-BE/assets/59831262/bb350825-c029-431f-985f-a9427bd28f54)
위와 같이 aws에 올려준 결과 아래와 같이 secrets를 사용할 수 있었습니다.
**application.yml**
```
  config:
    import:
      - 'aws-secretsmanager:dev/mykkumi/s3'
    s3:
      bucket: ${s3_bucket_name}
    credentials:
      access-key: ${s3_access_key}
      secret-key: ${s3_secret_key}
```
### 4. aws secrets manager로 인한 ci에서의 에러
develop 브랜치에 다시 Merge해봐야겠지만 ci 과정 중 에러가 발생했습니다. 
아래와 같은 에러가 발생하는 것으로 확인됩니다. 
```
java.lang.IllegalStateException at DefaultCacheAwareContextLoaderDelegate.java:180
        Caused by: org.springframework.boot.context.config.ConfigDataResourceNotFoundException at ConfigDataResourceNotFoundException.java:97
            Caused by: io.awspring.cloud.secretsmanager.AwsSecretsManagerPropertySources$AwsSecretsManagerPropertySourceNotFoundException at AwsSecretsManagerPropertySources.java:93
                Caused by: com.amazonaws.SdkClientException at AWSCredentialsProviderChain.java:136
6 actionable tasks: 6 executed
FAILURE: Build failed with an exception.
```
코드 자체에 문제가 있나 싶어 파일을 scp로 직접 서버로 옮겨 실행시켜봤으나, 수동으로 옮겨 실행했을 때는 잘 실행되는 것을 확인했습니다. 원인을 잘 모르겠어서 일단 수동 배포해놓고 에러 원인을 찾아보며 고쳐보려고 합니다.

eekrwl committed:
KKUMI-16 [FEAT] db 연결 및 aws secrets manager

AWS RDS를 통해 mysql DB를 연결했다.
일단 create로 계속 생성되도록 만들었는데 추후 수정할 예정이다.
RDS secrets 관리는 AWS의 Secrets Manager를 사용했다.

eekrwl committed:
KKUMI-16 [FEAT] Banner 테이블 추가, 리포지토리 생성

eekrwl committed:
KKUMI-16 [FEAT] banners 전체 호출 api

홈화면 상단에 들어갈 배너 전체 호출 api를 만들었습니다.

eekrwl committed:
KKUMI-16 [UPDATE] DB create->update 설정 변경

data.sql을 사용하여 실행시마다 sql쿼리를 실행해서 데이터를 넣으려다가
CI/CD에도 변경이 필요할듯하여 update 사용하기로 변경

eekrwl committed:
KKUMI-16 [CHORE] 변수명 형식 변경

변경전: snake case
변경후: camel case

eekrwl committed:
KKUMI-16 [CHORE] 필요없는 주석 제거

eekrwl committed:
KKUMI-16 [DOCS] banners api swagger 설명 추가

eekrwl committed:
KKUMI-16 [UPDATE] banner에 배너이미지 외 상세 페이지 추가

홈화면 상단에서 보여줄 banner 외에 클릭했을 때 보여줄 상세 이미지 추가

eekrwl committed:
KKUMI-16 [FIX] 배너 전체 호출 api 수정

변경전: banner에 imageUrl만 존재
변경후: banner에 imageUrl뿐만 아니라 detailImageUrl도 존재
해당 부분 반영하여 코드 수정했습니다.

eekrwl committed:
KKUMI-16 [FEAT] 배너 id로 상세배너 호출 api 구현

babjo commented:
> 하지만 생각해보니 홈화면에서 보이는 배너는 요약된 대표화면이고, 클릭을 했을때 보이는 상세 이미지가 없다는 생각이 들었습니다. 그래>서 아래와 같이 detailImageUrl을 추가했습니다.

클릭했을 때 나오는 페이지가 별도로 있나요? 와이어프레임상에 있을까요?

> ErrorCode에 BANNER_NOT_FOUND라고 넣어놨는데, 크기를 너무 작게 잡아 재사용할 일이 없는 예외를 만든건지 고민입니다. 그냥 해당 자료가 없을 때, ErrorCode에는 NOT_FOUND 정도만 넣어놓고, 상세 설명에서 배너가 존재하지 않는다고 적어주는 것으로 바꿀지 고민입니다.

BANNER_NOT_FOUND errorcode 를 클라이언트가 받았을 때 특정해야하는 동작이 있을까요? 애초에 거의 일어나지 않을 것이며 NOT_FOUND 해도 무방할거같습니다.

> 원인을 잘 모르겠어서 일단 수동 배포해놓고 에러 원인을 찾아보며 고쳐보려고 합니다.

넵, 수정이 어렵다면 해당 repo 가 private 이니 외부 노출 위험은 없으니 secrets manager 사용하지 않고 github 에 저장해도 되지 않을까합니다.

babjo commented on src/main/java/com/swmarastro/mykkumiserver/domain/banner/BannerController.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @GetMapping("/{id}")
+    @Operation(summary = "배너 id로 배너 상세 불러오기", description = "배너 id로 배너 상세 이미지를 불러옵니다.")
```
상세 불러오기 기능 관련 와이어프레임이 있을까요?

eekrwl replied:
> 상세 불러오기 기능 관련 와이어프레임이 있을까요?

와이어프레임에 빠져있었습니다..!!🥹 개발문서 api 목록에만 존재해서 새로 만들었습니다

eekrwl commented:
> 클릭했을 때 나오는 페이지가 별도로 있나요? 와이어프레임상에 있을까요?

와이어프레임에는 빠져있고 api 문서에만 있었습니다..!! 추가했습니다!

> BANNER_NOT_FOUND errorcode 를 클라이언트가 받았을 때 특정해야하는 동작이 있을까요? 애초에 거의 일어나지 않을 것이며 NOT_FOUND 해도 무방할거같습니다.

넵 NOT_FOUND로 수정하겠습니다!

> 넵, 수정이 어렵다면 해당 repo 가 private 이니 외부 노출 위험은 없으니 secrets manager 사용하지 않고 github 에 저장해도 되지 않을까합니다.

레포지토리가 public으로 올라가있습니다🥹 merge해서 한번 시도해보고 안된다면 기존 .yml에 저장하는 방법으로 돌려보겠습니다!
cicd 수행중 test를 할 때 문제가 생기는 것 같아 일단 test 제외해두었습니다. ㅎㅎ..

eekrwl committed:
KKUMI-16 [CHORE] banner_not_found -> not_found 에러 코드 이름 변경

BANNER_NOT_FOUND 에러코드는 재사용할 일이 없을 것 같아,
NOT_FOUND로 변경해주었습니다.
상세한 에러에 관해서는 message와 detail을 통해 각 상황에 따라 전달할 예정입니다.

babjo commented:
> 와이어프레임에는 빠져있고 api 문서에만 있었습니다..!! 추가했습니다!

배너를 보내는 상세페이지가 있다는 말씀이시죠? 배너 하나만 보는??
<img width="684" alt="image" src="https://github.com/SW-Marastro/Mykkumi-BE/assets/5077919/8af985d2-4f51-4cfe-992f-20528465df15">

> 레포지토리가 public으로 올라가있습니다🥹 merge해서 한번 시도해보고 안된다면 기존 .yml에 저장하는 방법으로 돌려보겠습니다!
cicd 수행중 test를 할 때 문제가 생기는 것 같아 일단 test 제외해두었습니다. ㅎㅎ..

헉 repo 는 private 으로 만들어주셔야할거같습니다.

eekrwl commented:
> 배너를 보내는 상세페이지가 있다는 말씀이시죠? 배너 하나만 보는??

네 피그마에도 빠져있어서 이렇게 간단하게 하나 만들어놨습니다!
<img width="241" alt="배너상세" src="https://github.com/SW-Marastro/Mykkumi-BE/assets/59831262/2634b85f-de3c-4d59-8565-dbf05a288710">

> 헉 repo 는 private 으로 만들어주셔야할거같습니다.

나중에 public 전환해야하지 않을까해서 일단 public으로 해두었는데 key같은게 보이는걸 방지하려고 보통 private으로 하나요..??

babjo commented:
> 보통 private으로 하나요..??

넵 나중에 public 으로 두는건 포트폴리오 때문인거죠?

babjo commented:
> 네 피그마에도 빠져있어서 이렇게 간단하게 하나 만들어놨습니다!

넵 이해했습니다.
